RFC 005 - Template contributing proposed process (#10)
thegooddocsproject/request-for-comment: template-contributing-guide.md

We value all contributions to the Good Docs Project initiatives, including templates and our other initiatives. The Join the community section of this guide explains how to get involved in those other related initiatives and provides links for more information.

## Table of contents

- [Before you start](#before-you-start)
  - [Template roles and resources](#template-roles-and-resources)
  - [Required deliverables](#required-deliverables)
- [Overview of the template writing phases](#overview-of-the-template-writing-phases)
- [Join the community](#join-the-community)
- [Adopt a template](#adopt-a-template)
  - [Guidelines for choosing a template](#guidelines-for-choosing-a-template)
- [Research and draft the template](#research-and-draft-the-template)
  - [Recommended research strategies](#recommended-research-strategies)
- [Workshop the draft of the template](#workshop-the-draft-of-the-template)
  - [Accepting feedback from others](#accepting-feedback-from-others)
- [Submit a pull request](#submit-a-pull-request)
- [Improve the template with user feedback](#improve-the-template-with-user-feedback)

## Before you start

> Before starting, ensure you are familiar with The Good Docs Project’s goals, key concepts, and supported personas. For more information, see [About the Good Docs Project](https://thegooddocsproject.dev/about/).

Reviewer: This link does not appear to have any information about personas. If we are arguing that someone needs to be familiar with our personas before reading this, we should link directly to info about the personas. I'm not convinced, however, that knowing about our personas is a requirement for reading through the contributing process. That's probably something that can be introduced to new members in the writing working groups. Or we can link to it as part of the contributing guide.

Reviewer 2: I don't think we need to say anything about personas at this point. I'm not missing any context by not knowing about personas.

Author: Fair enough. I'll remove "personas."

Author: I've adopted this suggestion and I'm resolving this comment.

- **The templates coordinator** - This templateer oversees our overall template development process as a project manager and provides assistance to contributors working on templates.
- **Template mentors** - This group of experienced templateers lead template writing working groups where you can receive guidance and mentorship while working on your template.
- **Community reviewers** - This group of templateers are available to review templates during the workshop phase. They include members of your template workshop but also members from the larger Good Docs Project community who have volunteered to help review templates.
- **Pull request reviewers** - This group of experienced templateers review and approve pull requests submitted to the templates repository.

In addition to these roles, these resources are available to help you as you contribute templates to our project:

- **Working groups** - Our project is organized into several working groups that meet on a regular basis to work on the project’s key initiatives. To effectively contribute to a project, consider joining one of these working groups.

> - **Templates issue list** - All the templates that are actively being worked on or which could be worked on have a corresponding issue in the templates repository. Use the issue list or the kanban board to find a template to work on and track your template’s progress as it moves throughout the template writing phases.

Reviewer: It would be good to mention that you can find all the template writing issues by sorting on the "templates" label.

Author: We'll have to revise this document after the fact as our project management tools and processes change.

These roles and resources (including the names of the individuals currently serving in these roles) are explained in the [Template Roles and Resources](template-roles-and-resources.md) guide.

### Required deliverables

Currently, for a template set to be considered complete, each template set should have these documentation deliverables:

<table>
  <tr>
    <th>Deliverable</th>
    <th>Description and purpose</th>
    <th>Required?</th>
  </tr>
  <tr>

>     <td>Raw template</td>

Reviewer: Raw template does not seem like the right term for this file in the template set. Why not just call it "Template" or "Template File"? Raw implies there is some sort of refinement that has not happened yet. But this file will still be included in the template set after it has undergone all community feedback and revisions. In what sense is it "raw" at that point?

Author: Sure, that was just the wording used in the base template, but I'm fine with just "template" or "template file."

Author: I've adopted this suggestion. Resolving the comment.

    <td>The raw template is the basic template for the type of document you are creating. It provides a rough outline of the suggested content and a few embedded writing tips for how to fill in the different sections of the template.</td>
    <td>Yes, required</td>
  </tr>
  <tr>
    <td>Template guide</td>

>     <td>This guide provides a deeper explanation of how to fill in the raw template. It provides a lightweight introduction to the purpose of this documentation and explains how to fill in each section of the document. Additional content may be included as needed.</td>

Reviewer: If anything should be called "Before you start", shouldn't it be the guide? I actually think guide is a good name, but note from the description here that this is the essential info the user needs before they fill in the template.

Author: See my reply to your previous comment.

Author: I've updated our document based on what we agreed to in last week's US general meeting. Resolving.

    <td>Yes, required</td>
  </tr>
  <tr>
    <td>Before you start guide</td>

>     <td>This optional guide is designed to help template users understand some of the key research, brainstorming, or pre-writing steps that they might need to do before adding this type of document to a guide. If this documentation type has certain key decisions that need to be made with stakeholders or if it has a dependency on some other type of document or process, that should be mentioned in this guide.</td>

Reviewer: If this info is essential to using the template, it should be in ***-guide. I don't see a coherent explanation here distinguishing what goes in guide vs. what goes in **-before-you-start. There seems like considerable overlap to me..

Author: See my reply to your previous comment.

Author: I've updated our document based on what we agreed to in last week's US general meeting. Resolving.

    <td>No, optional</td>
  </tr>
  <tr>
    <td>Template example</td>
    <td>The Good Docs Project is in the process of designing a fake documentation project that can provide examples of our templates in action. Until this project is ready, it is optional to provide an example of the template. However, if you want to create a fake example as part of your template set, you may do so.</td>

>     <td>No, optional</td>

Reviewer 2: I personally think this should be required. It's literally my best resource when I'm looking at templates, because of the way my brain works. It should be a trivial ask.

Author: Unfortunately, it has to be optional since our example project is not available yet.

Author: I've noted this comment in the feedback section of the RFC for now.

  </tr>
</table>

For more detailed information about each deliverable, see [Template Deliverables](template-deliverables.md).

## Overview of the template writing phases
Contributing a template project to our repository has five phases, as summarized in this table:

<table>
  <tr>
    <th>Phase</th>
    <th>Your goals</th>
    <th>Resources to help you</th>
  </tr>
  <tr>
    <td>1 - Join the community</td>
    <td><ul>
          <li>Join our project’s communication channels.</li>
          <li>Decide which initiative you’d like to work on based on your interests and experience.</li>
        </ul>
    </td>
    <td><ul>
          <li>The Good Docs Project welcoming committee</li>
          <li>Working groups</li>
        </ul>
    </td>
  </tr>
  <tr>
    <td>2 - Adopt a template</td>
    <td><ul>
          <li>Join a template writing working group led by a template mentor.</li>
          <li>Work with a template mentor and your working group to decide which template you’ll work on.</li>

>           <li>Assign yourself to the corresponding issue for that template.</li>

Reviewer 3: As it stands right now I'd expect templates to be represented by epics, either release-specific or otherwise. We may need to adjust this wording accordingly.

Author: Agreed. Do we need to resolve this before the RFC is adopted or can this be revised after the fact?

Author: I'm not going to take action on this comment because it's out of the scope of this RFC. Creating the template project management system will be a separate RFC. I did note this in the feedback section of the RFC, though.

        </ul>
    </td>
    <td><ul>
          <li>The issues board on the templates repo</li>
          <li>Template mentors</li>
          <li>Templates coordinator</li>
          <li>Template working groups</li>
        </ul>
    </td>
  </tr>
  <tr>
    <td>3 - Research and draft</td>
    <td><ul>
          <li>Research examples and best practices for the type of template you are working on.</li>
          <li>Collaborate and get early feedback on your research from your template mentor and/or other templateers as part of a template writing working group.</li>
          <li>Create drafts of the template set deliverables in Google Docs or your preferred tool.</li>
        </ul>
    </td>
    <td><ul>
          <li>Working groups</li>
          <li>Template mentors</li>
          <li>Templates coordinator</li>
        </ul>
    </td>
  </tr>
  <tr>

>     <td>4 - Workshop the drafts with the community</td>

Reviewer: I'd change this to "Revise the drafts based on community review"

Author: See my reply to your previous comment.

Author: I'm adopting your suggestion, per our discussion in last week's general meeting. Resolved: We had a lengthy discussion about it at our last general meeting. The consensus was that the term "workshop," although accurate, was not necessarily understood across all cultures. For that reason, we're going to change the name of this phase to "Get feedback on drafts from the community." There was also some confusion about who could review these drafts. We'll clarify that community reviewers includes anyone who is a member of the Good Docs Project, including members of template writing working groups. For now, we'll keep the requirement to have at least three community reviews are required to move to the next phase. We'll adjust this requirement if we discover it is too difficult.

    <td><ul>

>           <li>Revise and refine the drafts of your deliverables with help from at least three community reviewers.</li>

Reviewer: Since this is a requirement, I'd make it more clear what is required. Three community members must review the draft and provide feedback. The templateer should revise their draft based on the feedback they receive (if any).

Author: Agreed.

Author: I've adopted this suggestion and have made revisions to this part of the guide. Resolving.

          <li>Work with your template mentor and the templates coordinator to determine when your draft is ready for the next phase.</li>
        </ul>
    </td>
    <td><ul>
          <li>Community reviewers</li>
          <li>Subject matter experts and members of other technical writing communities</li>
          <li>Template mentors</li>
          <li>Templates coordinator</li>
        </ul>
    </td>
  </tr>
  <tr>
    <td>5 - Submit a pull request</td>
    <td><ul>
          <li>If needed, convert drafts from Google Docs or your preferred tool to Markdown.</li>
          <li>Open a pull request. NOTE: If you are unfamiliar with Git and GitHub, your template mentor or another community member can help you.</li>
          <li>Revise documents based on requests from pull request reviewers.</li>
        </ul>
    </td>
    <td><ul>
          <li>Pull request reviewers</li>
          <li>Template mentors</li>
          <li>Templates coordinator</li>
        </ul>
    </td>
  </tr>
  <tr>
    <td>6 - Improve the template with user feedback</td>
    <td><ul>

>           <li>After completing the previous phase, your template will officially be part of the Good Docs Project and will be available to our users.</li>

Reviewer: We might want to add a requirement here to create an issue to handle feedback and do the next version of the template, after it is finalized. This can then be assigned directly to the creator, or left unassigned (surfacing clearly that there is a gap in coverage, to facilitate governance of the content). But perhaps the template coordinator or mentors could do that.

Author: Let's leave this as an open action item that needs to be resolved when we work out the finer details of the template project management process.

Author: I've added your comment to the feedback section of the RFC.

As you begin to join our project, remember that this is a project composed entirely of volunteers.
We love to welcome new members, but want to be careful not to burn out our core project contributors.
This level of mindfulness helps us ensure that we retain our project’s capacity to produce high-quality work.
As such, we ask that you respect the time of our project maintainers and contributors.
(And expect us to respect your time in return!)

We expect all members of our project to be nice to each other and to follow our [Code of Conduct](https://thegooddocsproject.dev/code-of-conduct/) when interacting with other members of the Good Docs Project.

## Adopt a template

In this phase, you will decide which template project you will work on and adopt that template.

Be aware that:

> - Each template project is represented by a corresponding issue in the [Good Docs Project templates](https://github.com/thegooddocsproject/templates/issues) repository.

Reviewer 3: Issue or epic, TBD.

Author: See my reply to your previous comment.

> - Each template project is represented by a corresponding issue in the [Good Docs Project templates](https://github.com/thegooddocsproject/templates/issues) repository.

Reviewer: Good place to mention the `templates` label.

Author: We'll have to revise this document after the fact as our project management tools and processes change.

- You will use this issue to communicate the status of your template project as it moves through the different phases of the template writing process.
- The Good Docs Project managers use a kanban board that shows all the issues for the current template projects. This tool allows the templates coordinator, template mentors, community reviewers, and other stakeholders to track the overall progress of each template and assist templateers whose progress is blocked.

To adopt a template:

1. Scroll through the list of template issues and see if one interests you and/or matches your skill set. Alternatively, if you have an idea for a template project that does not yet have an issue, and you have the support of a template mentor, you can create a new issue for your template project.

> 2. Assign yourself to the issue. See Assigning issues and pull requests in the GitHub docs for more information.

Reviewer: Wouldn't hurt to add the link: https://docs.github.com/en/issues/tracking-your-work-with-issues/assigning-issues-and-pull-requests-to-other-github-users

Author: Good call. I'll add that.

Author: I've adopted this suggestion. Resolving.

- **Searching for guides, books, blog posts, conference presentations, or videos about best practices.** Search the Internet to find advice, tips, or expert research about how to create that type of document. Consider posting in a forum for resource ideas. For example, asking for helpful guides or insights on a community forum like the Write the Docs Slack workspace could be beneficial. Be mindful of, and respect copyright terms of source material. Do not plagiarize and offer attribution where appropriate.
- **Reaching out to experts.** When you find people you admire, who have researched your topic already, try reaching out to them. They often have a “how to contact me” webpage. Ask if they’d be okay with their material being used. (They might need to republish under a different copyright.) Invite them to participate in the template working group. They might even be prepared to lead it. If you feel shy about reaching out yourself, your template mentor or senior Good Docs Project member might offer to help.

> - **Collaborating with others in a working group.** Consider joining one of the template-focused working groups to meet regularly with a workshop of writers and discuss research ideas and findings.

Reviewer: The phrasing here is "consider". But elsewhere it seemed more like a hard requirement that template authors join a template-writing working group. We should try to make more clear whether or not that is a hard requirement.

Author: It's a hard requirement and this reflects that our thinking about making it a requirement shifted over the course of drafting this guide. I'll update this to communicate that it's a hard requirement.

Author: I've adopted this suggestion.

## Workshop the draft of the template

In this phase, you’ll begin to share your draft with community reviewers and invite feedback.
Optionally, you might also consider sharing it beyond our community with other technical writing communities such as Write the Docs or beyond.

> The workshop phase is arguably the most crucial and important phase in the template writing process, so your template project might spend the bulk of its time in this phase.

Reviewer: We might want to rephrase this to avoid the term "to workshop". Perhaps call this the "feedback phase" or the "review phase".

Author: I don't like "feedback" because it isn't a verb the way the other phases are. "Review" could work though.

Author: Yes, we've reached a decision about this in the US general meeting last week. Resolved: We had a lengthy discussion about it at our last general meeting. The consensus was that the term "workshop," although accurate, was not necessarily understood across all cultures. For that reason, we're going to change the name of this phase to "Get feedback on drafts from the community." There was also some confusion about who could review these drafts. We'll clarify that community reviewers includes anyone who is a member of the Good Docs Project, including members of template writing working groups. For now, we'll keep the requirement to have at least three community reviews are required to move to the next phase. We'll adjust this requirement if we discover it is too difficult.

To share your Google Docs drafts:

1. Inside the draft, click the **Share** button and change the **Get Link** settings to: **Anyone on the internet with this link can create comments.**

2. Copy the link to your Google Doc drafts into the issue that corresponds with your template in the templates repository.

3. Notify the templates coordinator or your template mentor, who will help you get reviews for your draft from the community reviewers group.

> The templates coordinator will help ensure you get at least three high quality reviews from the community reviewers. When you’ve received sufficient community input and incorporated suggestions into your draft, you will collaborate with a templates coordinator to determine that your draft is ready to move to the next phase.

Reviewer 3: I'm a little worried about the level of coordination that might be required here, especially when considering iterations on templates post-v1, where each minor change requested might be represented by an issue, each of which would need the required three reviews.

Reviewer: Maybe two reviewers needed for a new template, and one reviewer needed for an update to an existing template?

Author: Keep in mind that we already have a requirement to author templates as part of a template-focused working group. So if you could even just get three community reviewers from your group, that satisfies the requirement. Our two template-focused working group currently both have well above three members. And then you always have the larger community to draw upon, of which we have many more than three.

Author: In last week's general meeting, we decided that, for now, we'll keep the requirement to have at least three community reviews are required to move to the next phase. We'll adjust this requirement if we discover it is too difficult.

Reviewer 4: Non blocking comment: My memory of this meeting was that we wouldn't make "three" a hard requirement, but rather a strong recommendation.
```suggestion
The templates coordinator will help you find reviewers. Typically three or more community reviewers will provide a detailed review of your material. When you’ve received sufficient community input and incorporated suggestions into your draft, you will collaborate with a templates coordinator to determine that your draft is ready to move to the next phase.
```

Remember to be positive and show appreciation when people take time to review your drafts.
Providing feedback takes time and energy.
Treat each piece of feedback as a gift (even feedback that you possibly choose to disregard).
Happy editing!

## Submit a pull request

The purpose of this phase is to ensure your template project meets the standards of the Good Docs Project and is ready for public distribution.
In this phase, you’ll convert your template documents into Markdown and open a pull request in the `templates` repository on GitHub.

If you are not comfortable working in Markdown, Git, or GitHub, a community member (such as your template mentor) will assist you.

Once you’ve submitted a pull request, the templates coordinator will assign at least two members from the approved pull request reviewer list to review your template.
These reviewers will check the template to ensure it meets the project’s formatting and quality guidelines as outlined in our template pull request checklist.
Once the template has two approvals, it can be merged into the final project by a templates coordinator.

## Improve the template with user feedback

> After you’ve submitted your pull request in the previous phase, your template will officially be part of the Good Docs Project. Yay!

Reviewer: Is this true? Or are they officially part of the Good Docs Project after the PR is approved, rather than merely submitted?

Reviewer 2: The way it's described is as if the pull request is a formality. Should we smash in the reviews here? 

I think the reason that it's organized this way is because it takes work to get the template from the googly doc to the pull request, so we don't want people to do the work, then have to go back to the drawing board, change it, and then update in the PR?

Author: Well, it's when it's merged, so I'll clarify that.

We want to give people the option to do a Google Doc (because it's easier to review) or markdown. It's up to the template author.

Author: I've adjusted some of the text here to add more clarification that this is a final quality check with experienced project leaders doing the final review. Resolving this comment thread.

          <li>Work with a template mentor and your working group to decide which template you’ll work on.</li>
          <li>Assign yourself to the corresponding issue for that template.</li>
        </ul>
    </td>
    <td><ul>
          <li>The issues board on the templates repo</li>
          <li>Template mentors</li>
          <li>Templates coordinator</li>
          <li>Template working groups</li>
        </ul>
    </td>
  </tr>
  <tr>
    <td>3 - Research and draft</td>
    <td><ul>
          <li>Research examples and best practices for the type of template you are working on.</li>
          <li>Collaborate and get early feedback on your research from your template mentor and/or other templateers as part of a template writing working group.</li>
          <li>Create drafts of the template set deliverables in Google Docs or your preferred tool.</li>
        </ul>
    </td>
    <td><ul>
          <li>Working groups</li>
          <li>Template mentors</li>
          <li>Templates coordinator</li>
        </ul>
    </td>
  </tr>
  <tr>
    <td>4 - Get feedback on drafts from the community</td>
    <td><ul>

>           <li>Revise and refine the drafts of your deliverables based on feedback from at least three community reviewers. Community reviewers are anyone who is a member of the Good Docs Project, including members of template writing working groups.</li>

Reviewer 4: Suggestion: non-blocking: My recollection of our last meeting was that we have a "strong preference for three or more reviewers" and not make it a hard requirement.
```suggestion
          <li>Revise and refine the drafts of your deliverables. Expect to receive feedback from at least three community reviewers. Community reviewers are anyone who is a member of the Good Docs Project, including members of template writing working groups.</li>

```
